### PR TITLE
Add file and line to rule engine exception error messages

### PR DESCRIPTION
**Pull request by rleighton**
## Background

While testing rule errors in Snowflake I noticed that the errors were only of limited use because they included only the exception.
For example, take this rule:
```
def rule(event):
	# Return True to match the log event and trigger an alert.
	if len(event) > 0:
	    1/0
	    
	return False
```
Previously the error message would be:
```
ZeroDivisionError('division by zero')
```

This PR adds the rule file, the line number and the line:
```
division by zero: File "/tmp/rules/AlwaysFail.py", line 4, in rule 1/0
```

## Changes

- add method `error_message` to RuleResult to format the error logged with traceback info
- update tests

## Testing

- mage test:ci
- mage deploy, created a divide by zero rule and query results in Athena

### CODE REVIEW

File: internal/log_analysis/rules_engine/src/engine.py [hunk 1]
```diff
@@ -49,17 +49,21 @@ def analyze(self, log_type: str, event: Dict[str, Any]) -> List[EngineResult]:
             self.logger.debug('running rule [%s]', rule.rule_id)
             result = rule.run(event, batch_mode=True)
             if result.errored:
+                short_error_message = repr(result.rule_exception)
+                # TODO(kostaspap): remove error logging once error reporting notification system is in place
+                self.logger.error('failed to run rule %s %s', rule.rule_id, short_error_message)
+                error_type = type(result.rule_exception).__name__
```
Comment (rleighton) [inline internal/log_analysis/rules_engine/src/engine.py:55]: The `error_type` moved here from below is the same in Enterprise (this is a backport of testing done there).

File: internal/log_analysis/rules_engine/src/engine.py [hunk 2]
```diff
@@ -49,17 +49,21 @@ def analyze(self, log_type: str, event: Dict[str, Any]) -> List[EngineResult]:
             self.logger.debug('running rule [%s]', rule.rule_id)
             result = rule.run(event, batch_mode=True)
             if result.errored:
+                short_error_message = repr(result.rule_exception)
+                # TODO(kostaspap): remove error logging once error reporting notification system is in place
+                self.logger.error('failed to run rule %s %s', rule.rule_id, short_error_message)
```
Comment (kostaspap) [inline internal/log_analysis/rules_engine/src/engine.py:54]: You shouldn't log the error here - users will be notified through proper rule error alert

File: internal/log_analysis/rules_engine/src/rule.py [hunk 3]
```diff
@@ -59,6 +60,15 @@ class RuleResult:
     alert_context: Optional[str] = None
     alert_context_exception: Optional[Exception] = None
 
+    def error_message(self) -> Optional[str]:
+        """Returns formatted error message with traceback"""
+        if self.rule_exception is not None:
+            trace = traceback.format_tb(self.rule_exception.__traceback__)
+            # we only take last 2 elements of trace which will show the rule file name and line of the error, for example:
+            #    division by zero: File "/tmp/rules/AlwaysFail.py", line 4, in rule 1/0
```
Comment (kostaspap) [inline internal/log_analysis/rules_engine/src/rule.py:68]: You should just keep the file name, not the full file path. This is really Panther internal and doesn't provide any value to the users.

File: internal/log_analysis/rules_engine/tests/test_engine.py [hunk 4]
```diff
@@ -127,4 +127,10 @@ def test_analyse_many_rules_one_throws_exception(self) -> None:
             )
         ]
 
+        # since the error message will have a variable file path, test it separately
+        expected_event_matches[1].error_message = result[1].error_message
+        self.assertIsNotNone(result[1].error_message)
+        if result[1].error_message is not None:
+            self.assertTrue(result[1].error_message.endswith('''Exception("Found an issue")'''))
+
```
Comment (kostaspap) [inline internal/log_analysis/rules_engine/tests/test_engine.py:135]: See comment above - if you remove the full file path, it should be possible to update the check here as well.

File: internal/log_analysis/rules_engine/src/engine.py (hunk 2 above)
Comment (rleighton) [reply internal/log_analysis/rules_engine/src/engine.py:54]: Notice the `TODO`. It says you will do it :)

Is it ok now to remove?

File: internal/log_analysis/rules_engine/src/rule.py (hunk 3 above)
Comment (rleighton) [reply internal/log_analysis/rules_engine/src/rule.py:68]: Good idea, will do.

File: internal/log_analysis/rules_engine/src/engine.py (hunk 2 above)
Comment (kostaspap) [reply internal/log_analysis/rules_engine/src/engine.py:54]: It was removed on `master`. According to this PR, you are adding it back.

Comment (rleighton) [reply internal/log_analysis/rules_engine/src/engine.py]: Done

File: internal/log_analysis/rules_engine/src/rule.py (hunk 3 above)
Comment (rleighton) [reply internal/log_analysis/rules_engine/src/rule.py]: Done

File: internal/log_analysis/rules_engine/tests/test_engine.py (hunk 4 above)
Comment (rleighton) [reply internal/log_analysis/rules_engine/tests/test_engine.py]: Done

**Review by kostaspap — APPROVED** [review]

**Pull request merged**